python/ray/data/_internal/execution/operators/input_data_buffer.py

```diff
@@ -41,7 +42,8 @@ def __init__(
 
     def start(self, options: ExecutionOptions) -> None:
         if not self._is_input_initialized:
-            self._input_data = self._input_data_factory()
+            # Copy the input data to avoid mutating the original list.
+            self._input_data = copy.deepcopy(self._input_data_factory())
```

Comment:
Operators may mutate bundles, hence we need to deepcopy the original input data if taking multiple iterations over the dataset as streaming_split() supports.

Reply:
Can we avoid mutating the bundles instead? It'd be ideal if we can make the RefBundle object immutable and generate a new RefBundle object when needed. That will make our code less error-prone.

Reply:
Yeah, I think that is doable. At least, we can make the blocks a tuple field so callers don't pop from it unintentionally.

Reply:
Updated to make RefBundle blocks an immutable field.